File: parent/54 -- Spiral Matrix/mine2.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        # Fastest
        # Time: O(M×N)
        # Space: O(M×N)

        m, n = len(matrix), len(matrix[0])
        directions = [(0,1), (1,0), (0,-1), (-1,0)]
        res = []

        dirInd = 0
        visited = set()
        x,y = 0,0

        def isWithinBounds(x, y):
            return 0 <= x < m and 0 <= y < n and (x,y) not in visited

        while isWithinBounds(x,y):
            res.append(matrix[x][y])
            visited.add((x,y))

            new_x = x + directions[dirInd][0]
            new_y = y + directions[dirInd][1]

            if not isWithinBounds(new_x, new_y):
                # change direction
                dirInd = (dirInd + 1) % 4
                new_x = x + directions[dirInd][0]
                new_y = y + directions[dirInd][1]
            
            x, y = new_x, new_y
        
        return res

File: parent/54 -- Spiral Matrix/mine2.go
package main

func spiralOrder(matrix [][]int) []int {
	M, N := len(matrix), len(matrix[0])
	visited := make(map[[2]int]bool)
	res := []int{}
	directions := [4][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	x, y, dirInd := 0, 0, 0

	isInSet := func(set map[[2]int]bool, val [2]int) bool {
		_, exists := set[val]
		return exists
	}

	isOutOfBounds := func(x int, y int) bool {
		return 0 > x || x >= M || 0 > y || y >= N || isInSet(visited, [2]int{x, y})
	}

	for !isOutOfBounds(x, y) {
		res = append(res, matrix[x][y])
		visited[[2]int{x, y}] = true

		new_x := x + directions[dirInd][0]
		new_y := y + directions[dirInd][1]

		if isOutOfBounds(new_x, new_y) {
			dirInd = (dirInd + 1) % 4
			new_x = x + directions[dirInd][0]
			new_y = y + directions[dirInd][1]
		}

		x = new_x
		y = new_y
	}

	return res
}
